### ISSUE TITLE
'_SocketStreamRead' in CFSocketStream.c Does Not Reliably and Consistently Handle EAGAIN Following a New TCP Connection on Linux

### ISSUE THREAD

**Issue by gerickson** [gerickson/opencfnetwork#18]
Per [OpenHLX Issue 18](https://github.com/gerickson/openhlx/issues/18), the logic in `_SocketStreamRead` in _CFSocketStream.c_ does not always and reliably handle `EAGAIN` following a new TCP connection that has just indicated ready-for-read via `select` but does not yet have data for reading.

The deferred-create-private-run-loop-mode-and-poll seems, at least on Linux, unreliable and needlessly complex when the following is likely more than sufficient and, in fact, best practice on POSIX-based systems:

```
@@ -2825,7 +2825,13 @@ _CFSocketRecv(CFSocketRef s, UInt8* buffer, CFIndex length, CFStreamError* error
    
    else {      
        /* Try to read some bytes off the socket. */
-       result = read(CFSocketGetNative(s), buffer, length);
+#if defined(__linux__)
+        do {
+#endif
+           result = read(CFSocketGetNative(s), buffer, length);
+#if defined(__linux__)
+        } while ((result == -1) && (errno == EAGAIN));
+#endif
        
        /* If recv returned an error, get the error and make sure to return -1. */
        if (result < 0) {
```

and works 100% of the time.

As noted in [OpenHLX Issue 18](https://github.com/gerickson/openhlx/issues/18), the reason for the failure of the deferred-create-private-run-loop-mode-and-poll may be a deficiency in `__CFRunLoopRun` in CF itself under Linux; however, that path seems to lead to `ETIMEDOUT` failures rather than just recovering from the `EAGAIN`, reading the data when available, and successfully completing.

**Comment by gerickson** [comment]
> 

However, as others on the 'net have pointed out, the simple solution for sockets potentially puts the system in an undesirable indefinite poll loop on `read` in corner cases.

**Comment by gerickson** [comment]
There is a possibility that [CF-1153.18/CFSocket.c](https://opensource.apple.com/source/CF/CF-1153.18/CFSocket.c.auto.html) has a fix for this given the ineffectiveness of the `EAGAIN` approach in `_SocketStreamRead`. See the mention of "rdar://problem/17849895".

**Comment by gerickson** [comment]
Integrating fixes in CFSocket.c all the way up through CF-1153.18 alone do not appear to address this issue, at least on Linux.

At this point, the best approach (at least on Linux) appears to be taking a poll-on _CFSocketRecv with a time-bound loop until either `EAGAIN` clears or the time-bound of that loop expires. Across several tens of thousands of tests, this approach seems consistent in matching Linux behavior with iOS/macOS/tvOS behavior.

The diff amounts to:

```
@@ -828,8 +855,50 @@ _SocketStreamRead(CFReadStreamRef stream, UInt8* buffer, CFIndex bufferLength,
        }
        
        /* If there's no error, try to read now. */
+
+        /* Note that in certain cases, at least with TCP sockets,
+        ** there may be a small single-digit to mid double-digit
+        ** millisecond window of time between when the socket is
+        ** connecting and the select/poll signal that the socket is ready
+        ** for reading and when the first data shows up on the socket
+        ** for actual reading. In that intervening time, read attempts
+        ** on a non-blocking socket will return EAGAIN.
+        **
+        ** At least on Linux, experience has shown that the
+        ** EAGAIN-memset-continue and subsequent
+        ** _kCFStreamSocketReadPrivateMode approach found after this
+        ** block does not actually work to resolve this situation. The
+        ** socket either experiences a time out (which never happens
+        ** on iOS/macOS/tvOS) or the process incorporating the socket
+        ** stream deadlocks indefinitely for a retry that never
+        ** arrives (which likewise never happens on
+        ** iOS/macOS/tvOS). Consequently, given the short windows
+        ** involved and the seeming-ineffectiveness of the current
+        ** solution for this, we employ a similar time-bound (as above)
+        ** retry, continuing _CFSocketRecv, until the receive is
+        ** successful or until the retry timeout expires.
+        */
        else if (!ctxt->_error.error) {
-           result = _CFSocketRecv(ctxt->_socket, buffer, bufferLength, &ctxt->_error);
+            CFAbsoluteTime later;
+            CFTimeInterval interval;
+            CFNumberRef value = (CFNumberRef)CFDictionaryGetValue(ctxt->_properties, _kCFStreamPropertyReadTimeout);
+
+            if (!value || !CFNumberGetValue(value, kCFNumberDoubleType, &interval))
+                interval = kReadWriteTimeoutInterval;
+
+            later = (interval == 0.0) ? DBL_MAX : CFAbsoluteTimeGetCurrent() + interval;
+
+            do {
+                result = _CFSocketRecv(ctxt->_socket, buffer, bufferLength, &ctxt->_error);
+
+                if (result >= 0) {
+                    break;
+                }
+
+                if ((ctxt->_error.error != EAGAIN) && (ctxt->_error.domain == _kCFStreamErrorDomainNativeSockets)) {
+                    break;
+                }
+            } while (0 < (interval = (later - CFAbsoluteTimeGetCurrent())));
        }
        
        /* Did a read, so the event is no longer good. */
```

**Comment by gerickson** [comment]
Closed with #21.
